### ISSUE TITLE
ERROR: could not serialize access due to concurrent update when ingesting aspects with PostgreSQL database

### ISSUE THREAD

**Issue by Masterchen09** [datahub-project/datahub#8257]
I know that there is already a closed issue related to this bug (#7757), but I want start here from scratch in order to try to collect all available information to hopefully get this bug fixed in the future - I am also not sure if the closed issue will get any attention, when there are new comments.

Since the v0.10.1 release, when the default thread count of the datahub-rest sink was increased to 15 threads by default (#7675), we are facing an issue in relation to the PostgreSQL database when aspects are ingested into DataHub. Sometimes the ingestion works sometimes not, but somehow it seems to be related to the amount of aspects which are ingested (the more aspects the more likely it is, because of more concurrent transactions).

As mentioned in the title and the other issue the ingestion is failing with the message: "ERROR: could not serialize access due to concurrent update when ingesting aspects"

Currently there is a workaround available by simply reducing the thread count back to the old default of 1 thread (see the notice in the release notes for v.0.10.2 or [here](https://datahubspace.slack.com/archives/C029A3M079U/p1681763172957259?thread_ts=1680630315.514719&cid=C029A3M079U)), but based on the tests I have done so far, I have some doubts that the workaround solves all potential issues (it might work when only one ingestion is running at the same time, but I could also imaging this issus could also have influence when multiple users are changing aspects using the UI, unlikely but possible?).

Someone in Slack also mentioned that using the datahub-kafka sink instead of the datahub-rest sink solved the issue for them, but I think this only moves the error message away from the ingestion to the mce-consumer-job (standalone or built-in GMS should not make a difference here, except the location where the error message is logged?).

Although in the original issue already a log is available, here is a clean stack trace from the GMS when the error is occuring:

<details><summary>Stack Trace</summary>

```
com.linkedin.restli.server.RestLiServiceException: javax.persistence.PersistenceException: Error when batch flush on sql: update metadata_aspect_v2 set metadata=?, createdOn=?, createdBy=?, createdFor=?, systemmetadata=? where urn=? and aspect=? and version=?
at com.linkedin.metadata.restli.RestliUtil.toTask(RestliUtil.java:42)
at com.linkedin.metadata.restli.RestliUtil.toTask(RestliUtil.java:50)
at com.linkedin.metadata.resources.entity.AspectResource.ingestProposal(AspectResource.java:186)
at jdk.internal.reflect.GeneratedMethodAccessor630.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:566)
at com.linkedin.restli.internal.server.RestLiMethodInvoker.doInvoke(RestLiMethodInvoker.java:177)
at com.linkedin.restli.internal.server.RestLiMethodInvoker.invoke(RestLiMethodInvoker.java:333)
at com.linkedin.restli.internal.server.filter.FilterChainDispatcherImpl.onRequestSuccess(FilterChainDispatcherImpl.java:47)
at com.linkedin.restli.internal.server.filter.RestLiFilterChainIterator.onRequest(RestLiFilterChainIterator.java:86)
at com.linkedin.restli.internal.server.filter.RestLiFilterChainIterator.lambda$onRequest$0(RestLiFilterChainIterator.java:73)
at java.base/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:753)
at java.base/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:731)
at java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2108)
at com.linkedin.restli.internal.server.filter.RestLiFilterChainIterator.onRequest(RestLiFilterChainIterator.java:72)
at com.linkedin.restli.internal.server.filter.RestLiFilterChain.onRequest(RestLiFilterChain.java:55)
at com.linkedin.restli.server.BaseRestLiServer.handleResourceRequest(BaseRestLiServer.java:262)
at com.linkedin.restli.server.RestRestLiServer.handleResourceRequestWithRestLiResponse(RestRestLiServer.java:294)
at com.linkedin.restli.server.RestRestLiServer.handleResourceRequest(RestRestLiServer.java:262)
at com.linkedin.restli.server.RestRestLiServer.handleResourceRequest(RestRestLiServer.java:232)
at com.linkedin.restli.server.RestRestLiServer.doHandleRequest(RestRestLiServer.java:215)
at com.linkedin.restli.server.RestRestLiServer.handleRequest(RestRestLiServer.java:171)
at com.linkedin.restli.server.RestLiServer.handleRequest(RestLiServer.java:130)
at com.linkedin.restli.server.DelegatingTransportDispatcher.handleRestRequest(DelegatingTransportDispatcher.java:70)
at com.linkedin.r2.filter.transport.DispatcherRequestFilter.onRestRequest(DispatcherRequestFilter.java:70)
at com.linkedin.r2.filter.TimedRestFilter.onRestRequest(TimedRestFilter.java:76)
at com.linkedin.r2.filter.FilterChainIterator$FilterChainRestIterator.doOnRequest(FilterChainIterator.java:146)
at com.linkedin.r2.filter.FilterChainIterator$FilterChainRestIterator.doOnRequest(FilterChainIterator.java:132)
at com.linkedin.r2.filter.FilterChainIterator.onRequest(FilterChainIterator.java:62)
at com.linkedin.r2.filter.TimedNextFilter.onRequest(TimedNextFilter.java:55)
at com.linkedin.r2.filter.transport.ServerQueryTunnelFilter.onRestRequest(ServerQueryTunnelFilter.java:58)
at com.linkedin.r2.filter.TimedRestFilter.onRestRequest(TimedRestFilter.java:76)
at com.linkedin.r2.filter.FilterChainIterator$FilterChainRestIterator.doOnRequest(FilterChainIterator.java:146)
at com.linkedin.r2.filter.FilterChainIterator$FilterChainRestIterator.doOnRequest(FilterChainIterator.java:132)
at com.linkedin.r2.filter.FilterChainIterator.onRequest(FilterChainIterator.java:62)
at com.linkedin.r2.filter.TimedNextFilter.onRequest(TimedNextFilter.java:55)
at com.linkedin.r2.filter.message.rest.RestFilter.onRestRequest(RestFilter.java:50)
at com.linkedin.r2.filter.TimedRestFilter.onRestRequest(TimedRestFilter.java:76)
at com.linkedin.r2.filter.FilterChainIterator$FilterChainRestIterator.doOnRequest(FilterChainIterator.java:146)
at com.linkedin.r2.filter.FilterChainIterator$FilterChainRestIterator.doOnRequest(FilterChainIterator.java:132)
at com.linkedin.r2.filter.FilterChainIterator.onRequest(FilterChainIterator.java:62)
at com.linkedin.r2.filter.FilterChainImpl.onRestRequest(FilterChainImpl.java:106)
at com.linkedin.r2.filter.transport.FilterChainDispatcher.handleRestRequest(FilterChainDispatcher.java:75)
at com.linkedin.r2.util.finalizer.RequestFinalizerDispatcher.handleRestRequest(RequestFinalizerDispatcher.java:61)
at com.linkedin.r2.transport.http.server.HttpDispatcher.handleRequest(HttpDispatcher.java:101)
at com.linkedin.r2.transport.http.server.AbstractR2Servlet.service(AbstractR2Servlet.java:105)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
at com.linkedin.restli.server.RestliHandlerServlet.service(RestliHandlerServlet.java:21)
at com.linkedin.restli.server.RestliHandlerServlet.handleRequest(RestliHandlerServlet.java:26)
at org.springframework.web.context.support.HttpRequestHandlerServlet.service(HttpRequestHandlerServlet.java:73)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1631)
at com.datahub.auth.authentication.filter.AuthenticationFilter.doFilter(AuthenticationFilter.java:98)
at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:548)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
at org.eclipse.jetty.server.Server.handle(Server.java:516)
at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: javax.persistence.PersistenceException: Error when batch flush on sql: update metadata_aspect_v2 set metadata=?, createdOn=?, createdBy=?, createdFor=?, systemmetadata=? where urn=? and aspect=? and version=?
at io.ebean.config.dbplatform.SqlCodeTranslator.translate(SqlCodeTranslator.java:52)
at io.ebean.config.dbplatform.DatabasePlatform.translate(DatabasePlatform.java:219)
at io.ebeaninternal.server.transaction.TransactionManager.translate(TransactionManager.java:246)
at io.ebeaninternal.server.transaction.JdbcTransaction.translate(JdbcTransaction.java:698)
at io.ebeaninternal.server.transaction.JdbcTransaction.batchFlush(JdbcTransaction.java:680)
at io.ebeaninternal.server.transaction.JdbcTransaction.internalBatchFlush(JdbcTransaction.java:796)
at io.ebeaninternal.server.transaction.JdbcTransaction.flushCommitAndNotify(JdbcTransaction.java:1005)
at io.ebeaninternal.server.transaction.JdbcTransaction.commit(JdbcTransaction.java:1057)
at io.ebeaninternal.api.ScopeTrans.commitTransaction(ScopeTrans.java:136)
at io.ebeaninternal.api.ScopedTransaction.commit(ScopedTransaction.java:110)
at com.linkedin.metadata.entity.ebean.EbeanAspectDao.runInTransactionWithRetry(EbeanAspectDao.java:503)
at com.linkedin.metadata.entity.EntityService.ingestAspectToLocalDB(EntityService.java:573)
at com.linkedin.metadata.entity.EntityService.wrappedIngestAspectToLocalDB(EntityService.java:526)
at com.linkedin.metadata.entity.EntityService.upsertAspect(EntityService.java:980)
at com.linkedin.metadata.entity.EntityService.performUpsert(EntityService.java:929)
at com.linkedin.metadata.entity.EntityService.ingestProposal(EntityService.java:877)
at com.linkedin.metadata.resources.entity.AspectResource.lambda$ingestProposal$2(AspectResource.java:193)
at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
at com.linkedin.metadata.resources.entity.AspectResource.lambda$ingestProposal$3(AspectResource.java:193)
at com.linkedin.metadata.restli.RestliUtil.toTask(RestliUtil.java:30)
... 88 common frames omitted
Caused by: java.sql.BatchUpdateException: Batch entry 0 update metadata_aspect_v2 set metadata='xxx', createdOn='2023-06-15 17:46:49.37+00', createdBy='urn:li:corpuser:__datahub_system', createdFor=NULL, systemmetadata='{"registryVersion":"0.0.0.0-dev","runId":"66d85d13-bd9f-4c13-81dd-fdac3da292c4","registryName":"unknownRegistry","lastObserved":1686851209358}' where urn='urn:li:dataset:(urn:li:dataPlatform:xxx,PROD)' and aspect='xxx' and version=0 was aborted: ERROR: could not serialize access due to concurrent update Call getNextException to see other errors in the batch.
at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:165)
at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366)
at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:559)
at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:887)
at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:910)
at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1649)
at io.ebean.datasource.delegate.PreparedStatementDelegator.executeBatch(PreparedStatementDelegator.java:357)
at io.ebeaninternal.server.persist.BatchedPstmt.executeAndCheckRowCounts(BatchedPstmt.java:130)
at io.ebeaninternal.server.persist.BatchedPstmt.executeBatch(BatchedPstmt.java:97)
at io.ebeaninternal.server.persist.BatchedPstmtHolder.flush(BatchedPstmtHolder.java:124)
at io.ebeaninternal.server.persist.BatchControl.flushPstmtHolder(BatchControl.java:206)
at io.ebeaninternal.server.persist.BatchControl.executeNow(BatchControl.java:220)
at io.ebeaninternal.server.persist.BatchedBeanHolder.executeNow(BatchedBeanHolder.java:100)
at io.ebeaninternal.server.persist.BatchControl.flush(BatchControl.java:271)
at io.ebeaninternal.server.persist.BatchControl.flush(BatchControl.java:227)
at io.ebeaninternal.server.transaction.JdbcTransaction.batchFlush(JdbcTransaction.java:678)
... 103 common frames omitted
Caused by: org.postgresql.util.PSQLException: ERROR: could not serialize access due to concurrent update
at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
... 117 common frames omitted
```
</details>

Based on the stack trace I came across the [runInTransactionWithRetry](https://github.com/datahub-project/datahub/blob/d1ff9024b90829e00e1b1cb9666a7f7a683f6f10/metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanAspectDao.java#L493-L516) method in the class [EbeanAspectDao](https://github.com/datahub-project/datahub/blob/d1ff9024b90829e00e1b1cb9666a7f7a683f6f10/metadata-io/src/main/java/com/linkedin/metadata/entity/ebean/EbeanAspectDao.java). According to the current implementation, transactions are run with the stricter transaction isolation level REPEATABLE_READ (see also: #3285 resp. #3231) and transactions which fail with an RollbackException or DuplicateKeyException are retried.

However the exception which we are getting with the PostgreSQL database is none of them: The exception is a [PersistenceException](https://docs.oracle.com/javaee/7/api/javax/persistence/PersistenceException.html) (which is also the base class of the RollbackException and the DuplicateKeyException), therefore the transactions are not retried when this error is occuring.

Based on the information [here](https://pganalyze.com/docs/log-insights/app-errors/U138) the error message of the PostgreSQL database can happen when the REPEATABLE_READ isolation level is used and that the application should simply retry the transaction in such cases.

I have also tested the same scenario with a MySQL database, but I have added some logs to check whether transactions are retried and which exception was thrown in such a case and it seems that the same issue also occurs with the MySQL database, but as the transactions are retried it works with the MySQL database. With the MySQL database DuplicateKeyExceptions are thrown and when the transactions are retried it usually works.

I have implemented something similar to retry the transactions in case of the PostgreSQL database and this specific error message and when the transactions are retried it works too. Of course this is only another workaround yet and we might have to discuss when we should retry transactions here in general and of course how we can do it in case of the PostgreSQL database (checking the error message is definitely not really nice ;-)).

Here is the code I have used for testing:

```java
@Override
@Nonnull
public <T> T runInTransactionWithRetry(@Nonnull final Supplier<T> block, final int maxTransactionRetry) {
  validateConnection();
  int retryCount = 0;
  Exception lastException;

  T result = null;
  do {
    try (Transaction transaction = _server.beginTransaction(TxScope.requiresNew().setIsolation(TxIsolation.REPEATABLE_READ))) {
      transaction.setBatchMode(true);
      result = block.get();
      transaction.commit();
      lastException = null;
      break;
    } catch (RollbackException | DuplicateKeyException exception) {
      _logger.warn(exception.getClass().getCanonicalName() + " catched, will be retried...");

      lastException = exception;
    } catch (PersistenceException exception) {
      if (exception.getCause() instanceof BatchUpdateException) {
        Throwable cause = exception.getCause().getCause();

        if (cause != null) {
          if (cause.getMessage().equals("ERROR: could not serialize access due to concurrent update")) {
            _logger.error("ERROR: could not serialize access due to concurrent update catched, will be retried...");

            lastException = exception;
          } else {
            _logger.error(exception.getClass().getCanonicalName() + " catched, will not be retried (not the expected error message).");
            throw exception;
          }
        } else {
          _logger.error(exception.getClass().getCanonicalName() + " catched, will not be retried (cause of BatchUpdateException is null).");
          throw exception;
        }
      } else {
        _logger.error(exception.getClass().getCanonicalName() + " catched, will not be retried (cause of the exception is a " + cause.getClass().getCanonicalName() + " exception).");
        throw exception;
      }
    } catch (Exception exception) {
      Throwable cause = exception.getCause();
      if (cause != null) {
        _logger.error(exception.getClass().getCanonicalName() + " catched, will not be retried (cause of the exception is a " + cause.getClass().getCanonicalName() + " exception).");
      } else {
        _logger.error(exception.getClass().getCanonicalName() + " catched, will not be retried (cause of the exception is unknown).");
      }

      throw exception;
    }
  } while (++retryCount <= maxTransactionRetry);

  if (lastException != null) {
    throw new RetryLimitReached("Failed to add after " + maxTransactionRetry + " retries", lastException);
  }

  return result;
}
```

The documentation of the [PersistenceException](https://docs.oracle.com/javaee/7/api/javax/persistence/PersistenceException.html) mentions that the exceptions which inherit PersistenceException (and PersistenceException itself) except NoResultException, NonUniqueResultException, LockTimeoutException and QueryTimeoutException are indicating that the transaction is marked for rollback...maybe we should retry the transaction if it was marked rollback resp. was rollbacked?

FYI @david-leifker @thomasplarsson

**Comment by RyanHolstien** [comment]
Thanks for doing some digging here! The insights are definitely useful and appreciated. I think this addition probably makes sense. It's not a full solution, but I think it's still worth adding until we have something that fully solves it.

Retrying transactions makes sense for these situations I think, they are generally temporary collisions that get resolved. If you want to open up a PR for this we can get it reviewed.

**Comment by RyanHolstien** [comment]
Upgrading our Ebean version is on the backlog, the current version has several issues that cause pain including reporting all connection failures as NPEs. In the mean time, yes this seems to mesh with how Ebean has updated its error handling under the hood. Don't love that it's using a Spi, but if they left it open for usage then it should be fine until we update at which point we can use the updated Exception type.

Yeah I think filtering it to Postgres only for this code seems like the right decision for this change. Based on the linked commit the error code is different for Oracle and NuoDB (72000 and 40002 respectively) so I would suspect it's different for each platform.

**Comment by Masterchen09** [comment]
I have created a PR (#8278) with the proposed check of the SQL state in case of a PostgreSQL database.

With newer versions of io.ebean the platform seems to be directly available...I don't know if there is another possibility of doing it with the older version, but I have simply took the logic with the SpiServer from [here](https://github.com/ebean-orm/ebean/issues/2029), therefore I don't really think there is currently another possibility to do it (although it seems so, I am absolutely no Java expert nor I have done anything with io.ebean before this issue...I have simply used Google all the time 😁).

I have also checked whether the SQL state 40001 is returned by MySQL and it seems this is the case, but the SQL state means something different there (see [here](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-error-sqlstates.html)), therefore it is definitely right to check the platform to make sure it is a PostgreSQL database.
